### ISSUE TITLE
Mutation triggers unexpected refetch of related "cache-and-network" queries

### ISSUE THREAD

**Issue by zenflow** [apollographql/apollo-client#7938]
**Intended outcome:**

I have an app with Apollo Client which displays a list of "posts" on which users can "vote" (upvote/downvote).
The result of the `vote` mutation includes updates to post fields (`vote_total` & `my_vote_value`) which get merged into the normalized cache, so the update is reflected in the list of posts. This works great, since no extra requests are used to update the list of posts after the user "votes". I can even use the `optimisticResponse` option to optimize UX.

The problem is when I change the `fetchPolicy` for the `posts` query from default (`cache-first`) to `cache-and-network`.
I would expect this to cause the query to refetch **after the query is redisplayed** (which it does) but not any other random times.

```

**Actual outcome:**

After changing the `fetchPolicy` for the `posts` query to `cache-and-network`. the query is always refetched after a `vote` mutation.

Note that this happens even with the mutation's `refetchQueries` option set explicitly to `[]`.

**How to reproduce the issue:**

1. Create or find an AC app where a mutation response includes updates to a query shown on-screen
2. Change the fetch policy for the related query to `cache-and-network`
3. Start the app, open it, open the network tab in Chrome devtools
4. Trigger the mutation
5. Observe that the related query is refetched after the mutation finished

**Versions**
@apollo/client v3.3.13

**Comment by benjamn** [comment]
@zenflow A `FetchPolicy` like `cache-and-network` gets consistently (re)applied each time the cache is updated in a way that affects the query in question, so with `cache-and-network` you get a cache result _and_ a network result each time, per the policy.

I understand this is probably not the behavior you want, but rather than introducing ad-hoc new fetch policies like `cache-and-network-then-cache-first`, we introduced the `nextFetchPolicy` option, so you can sequence your fetch policies in a more flexible way: https://github.com/apollographql/apollo-client/issues/6760#issuecomment-668188727

If you use `fetchPolicy: "cache-and-network"` plus `nextFetchPolicy: "cache-first"`, your initial queries will receive both cache and network results, but subsequent updates to the cache will not trigger a network request unless the cache update _removed_ data relevant to the query, causing an incomplete cache read. I imagine that's closer to what you want?

**Comment by twalk4821** [comment]
@benjamn Thanks for the reply. This was surprising to us as well, as when we updated to client version 3 we started seeing extra requests for the same data (due to concurrent updates to fragments used by the main query). 

I had a follow up question as well, it's called `nextFetchPolicy` but it really means something like "all subsequent requests after the first one," correct?

**Comment by benjamn** [comment]
@twalk4821 That's right, `nextFetchPolicy` becomes the new `fetchPolicy` after the first request, until/unless you change it again.

**Comment by twalk4821** [comment]
@benjamn Oh I see -- every time a new request is made `nextFetchPolicy` will be evaluated again (not necessarily only after the first request)? I think it used to work that way where only the initial values passed to these properties would get evaluated but recently changed. Just not 100% sure what is the intended usage.

**Comment by twalk4821** [comment]
I think I answered my own question with this: 

https://github.com/apollographql/apollo-client/blob/da0a0e52b7b4b0743e9ee01c43e6adf1fa98553d/src/core/QueryManager.ts#L1018

**Comment by zenflow** [comment]
@benjamn Thank you for the very clear explanation!

I tried adding `nextFetchPolicy: "cache-only"` everywhere I am using `fetchPolicy: "cache-and-network"`, and that actually worked to stop AC from reloading `posts` after the `vote` mutation.. but the issue persists with another mutation!

Here is a new minimal example: (all queries using `fetchPolicy: "cache-and-network", nextFetchPolicy: "cache-only"`)
- a `session` query with a possible `user` object 
- a `posts` query that selects some `user` data through the `post.author` field
- a `submit_post` mutation that adds a post and selects the same `post` data including `post.author`

If the mutation function for `submit_post` is called __with an `update` option that prepends the result to the cached `posts` query__, then the `session` query is unexpectedly refetched after the mutation finishes. (Yes, even with the prescribed nextFetchPolicy) :(

**Comment by eltonio450** [comment]
Hello everyone,

Thanks for this

I feel the doc is not clear enough on this point. 

What I understood so far of "cache-and-network" : 
- given a query A, if enough data is present in the cache, a network call is made while data is served to the component A (I use React). When the updated data comes back from the server, the cache is updated, and subsequently the component
- if a concurrent query (B) updates the data of the cache used by query A, the data served to the component A is updated but no network call is triggered (I think it was working like this with Apollo Client 2)

What seems to be the current behavior :
- given a query A, if enough data is present in the cache, a network call is made while data is served to the component (I use React). When the updated data comes back from the server, the cache is updated, and subsequently the component
- if a concurrent query (B) updates the data of the first cache, the data served to the component is updated, and as the data has been updated, a network call is triggered for query A

What is not clear to me : 
- after the cache update by query B, is the network call triggered all the time or only if the data changed?
- if query A is mounted a second time, will it use the fetchPolicy or the nextFetchPolicy?
- in our case, query A triggers a refetch from query B, that triggers a refetch from query A, etc. Does it mean that we have a field that changes all the time?

Thank you!

**Comment by kanderek-mp** [comment]
I think I'm also seeing this.  We're on @apollo/client@npm:3.7.3. The original advice from @benjamn  to use `cache-only` does appear to work... but not when applying an optimistic response to the mutation which is updating the shared entity in the cache (albeit temporarily, I think). I'm not sure if it is expected but the cache-only fetch policy seems to cancel a network request (my expectation was that it would not need to make the request at all. There may be a detail I'm missing there.) for the query while the `cache-first` version will complete the request. This causes the response from the query to override the response from the mutation resulting in the UI being out of step. 

Anyways, for a fuller picture, my `fetchPolicy` was set to `networkOnly` and my nextFetchPolicy` was set to `cache-first`. I have a page with 1 query for SomeEntity and 1 mutation to update SomeEntity.  Strangely with these policies set and with an optimistic response for the mutation I see an additional network request for the query immediately after the update mutation is called _only some of the time_. I'm not sure I'm fully grasping the internals of Apollo Client 🥴.
